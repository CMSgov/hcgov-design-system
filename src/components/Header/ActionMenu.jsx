/* eslint-disable react/no-multi-comp, jsx-a11y/no-redundant-roles */
import { Button } from '@cmsgov/design-system';
import LoggedOutLinks from './LoggedOutLinks';
import PropTypes from 'prop-types';
import React from 'react';
import classnames from 'classnames';
import { translate } from 'react-i18next';

const menuId = 'hc-c-menu';

/**
 * ActionMenu is displayed at the top-right of the header.
 * There are two variations of this component: One for logged-out
 * users, which displays a "Log in" & "Espanol" link on desktop.
 * And another for logged-in users, which includes the user's first
 * name. The logged-in variation always displays a "Menu" toggle
 * button, and the logged-out variation only displays it on mobile.
 */
const ActionMenu = function (props) {
  return (
    <nav
      id="hc-c-header__actions"
      className="hc-c-header__actions ds-l-col ds-l-col--auto ds-u-margin-left--auto ds-u-font-weight--bold"
      role="navigation"
    >
      {props.loggedIn ? (
        <LoggedInActionMenu {...props} />
      ) : (
        <LoggedOutActionMenu {...props} />
      )}
    </nav>
  );
};

function LoggedInActionMenu(props) {
  return (
    <div>
      {props.firstName && (
        <span className="ds-u-display--none ds-u-sm-display--inline-block">
          {props.firstName}
        </span>
      )}
      <MenuButton {...props} />
    </div>
  );
}

function LoggedOutActionMenu(props) {
  return (
    <div>
      <LoggedOutLinks
        locale={props.locale}
        deConsumer={props.deConsumer}
        subpath={props.subpath}
        primaryDomain={props.primaryDomain}
        switchLocaleLink={props.switchLocaleLink}
      />
      <MenuButton
        {...props}
        className="ds-u-display--inline-block ds-u-sm-display--none"
      />
    </div>
  );
}

function MenuButton({ t, open, ...props }) {
  const className = classnames(
    'hc-c-action-menu-button',
    'ds-u-font-weight--bold',
    'ds-u-sm-margin-left--2',
    props.className
  );
  return (
    <Button
      inversed={props.inversed}
      aria-controls={menuId}
      aria-expanded={!!open}
      aria-label={open ? t('header.closeMenu') : t('header.openMenu')}
      className={className}
      onClick={props.onMenuToggleClick}
      size="small"
    >
      {open ? t('header.close') : t('header.menu')}
    </Button>
  );
}

ActionMenu.propTypes = {
  /** Applies the inverse theme styling */
  inversed: PropTypes.bool,
  firstName: PropTypes.string,
  locale: PropTypes.string.isRequired,
  loggedIn: PropTypes.bool,
  onMenuToggleClick: PropTypes.func.isRequired,
  /**
   * Indicates the menu is open, which influences the label
   * and ARIA attribute of the toggle button.
   */
  open: PropTypes.bool,
  /**
   * Indicates when a consumer is coming from a Direct Enrollment flow.
   * This will include additional messaging and modify some of the links.
   */
  deConsumer: PropTypes.bool,
  /**
   * [See `Header.props.subpath`]({{root}}/patterns/header/#patterns.header.react)
   */
  subpath: PropTypes.string,
};

LoggedInActionMenu.propTypes = {
  firstName: ActionMenu.propTypes.firstName,
  onMenuToggleClick: ActionMenu.propTypes.onMenuToggleClick,
  open: ActionMenu.propTypes.open,
};

LoggedOutActionMenu.propTypes = {
  onMenuToggleClick: ActionMenu.propTypes.onMenuToggleClick,
  open: ActionMenu.propTypes.open,
  deConsumer: ActionMenu.propTypes.deConsumer,
  locale: ActionMenu.propTypes.locale,
};

MenuButton.propTypes = {
  onMenuToggleClick: ActionMenu.propTypes.onMenuToggleClick,
  open: ActionMenu.propTypes.open,
  className: PropTypes.string,
};

export default translate()(ActionMenu);
