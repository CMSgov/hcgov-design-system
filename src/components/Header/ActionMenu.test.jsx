import ActionMenu from './ActionMenu';
import React from 'react';
import { mount } from 'enzyme';

describe('ActionMenu', function () {
  const handleMenuToggleClick = jest.fn();

  beforeEach(() => {
    handleMenuToggleClick.mockReset();
  });

  describe('logged-in version', () => {
    function component(props = {}) {
      return (
        <ActionMenu
          loggedIn
          firstName="John"
          onMenuToggleClick={handleMenuToggleClick}
          links={[]}
          {...props}
        />
      );
    }
    it('renders logged-in version', () => {
      expect(mount(component())).toMatchSnapshot();
    });

    it('set aria-expanded to true', () => {
      expect(mount(component({ open: true }))).toMatchSnapshot();
    });

    it('calls onMenuToggleClick', () => {
      const wrapper = mount(component());
      const button = wrapper.find('Button');

      button.simulate('click');

      expect(handleMenuToggleClick.mock.calls.length).toBe(1);
    });
  });

  describe('logged-out version', () => {
    /* eslint-disable react/no-multi-comp */
    function component(props = {}) {
      return (
        <ActionMenu
          loggedIn={false}
          locale="en"
          onMenuToggleClick={handleMenuToggleClick}
          links={[
            {
              label: 'label',
              href: 'href',
            },
          ]}
          {...props}
        />
      );
    }
    /* eslint-enable */

    it('renders logged-out version', () => {
      expect(mount(component())).toMatchSnapshot();
    });
  });
});
