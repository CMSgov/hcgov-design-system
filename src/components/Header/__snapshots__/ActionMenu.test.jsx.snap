// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionMenu logged-in version renders logged-in version 1`] = `
<Component
  firstName="John"
  links={Array []}
  loggedIn={true}
  onMenuToggleClick={[MockFunction]}
>
  <ActionMenu
    firstName="John"
    links={Array []}
    loggedIn={true}
    onMenuToggleClick={[MockFunction]}
    t={[Function]}
  >
    <nav
      className="hc-c-header__actions ds-l-col ds-l-col--auto ds-u-margin-left--auto ds-u-font-weight--bold"
      id="hc-c-header__actions"
      role="navigation"
    >
      <LoggedInActionMenu
        firstName="John"
        links={Array []}
        loggedIn={true}
        onMenuToggleClick={[MockFunction]}
        t={[Function]}
      >
        <div>
          <span
            className="ds-u-display--none ds-u-sm-display--inline-block"
          >
            John
          </span>
          <MenuButton
            firstName="John"
            links={Array []}
            loggedIn={true}
            onMenuToggleClick={[MockFunction]}
            t={[Function]}
          >
            <Button
              aria-controls="hc-c-menu"
              aria-expanded={false}
              aria-label="header.openMenu"
              className="hc-c-action-menu-button ds-u-sm-margin-left--2"
              component="button"
              onClick={[Function]}
              size="small"
              type="button"
            >
              <button
                aria-controls="hc-c-menu"
                aria-expanded={false}
                aria-label="header.openMenu"
                className="ds-c-button ds-c-button--small hc-c-action-menu-button ds-u-sm-margin-left--2"
                onClick={[Function]}
                type="button"
              >
                <MenuIcon
                  className="ds-u-margin-right--1"
                >
                  <svg
                    aria-hidden="true"
                    className="svg-inline--fa fa-bars fa-w-14 ds-u-margin-right--1"
                    data-icon="bars"
                    data-prefix="fas"
                    viewBox="0 0 448 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
                      fill="currentColor"
                    />
                  </svg>
                </MenuIcon>
                header.menu
              </button>
            </Button>
          </MenuButton>
        </div>
      </LoggedInActionMenu>
    </nav>
  </ActionMenu>
</Component>
`;

exports[`ActionMenu logged-in version set aria-expanded to true 1`] = `
<Component
  firstName="John"
  links={Array []}
  loggedIn={true}
  onMenuToggleClick={[MockFunction]}
  open={true}
>
  <ActionMenu
    firstName="John"
    links={Array []}
    loggedIn={true}
    onMenuToggleClick={[MockFunction]}
    open={true}
    t={[Function]}
  >
    <nav
      className="hc-c-header__actions ds-l-col ds-l-col--auto ds-u-margin-left--auto ds-u-font-weight--bold"
      id="hc-c-header__actions"
      role="navigation"
    >
      <LoggedInActionMenu
        firstName="John"
        links={Array []}
        loggedIn={true}
        onMenuToggleClick={[MockFunction]}
        open={true}
        t={[Function]}
      >
        <div>
          <span
            className="ds-u-display--none ds-u-sm-display--inline-block"
          >
            John
          </span>
          <MenuButton
            firstName="John"
            links={Array []}
            loggedIn={true}
            onMenuToggleClick={[MockFunction]}
            open={true}
            t={[Function]}
          >
            <Button
              aria-controls="hc-c-menu"
              aria-expanded={true}
              aria-label="header.closeMenu"
              className="hc-c-action-menu-button ds-u-sm-margin-left--2"
              component="button"
              onClick={[Function]}
              size="small"
              type="button"
            >
              <button
                aria-controls="hc-c-menu"
                aria-expanded={true}
                aria-label="header.closeMenu"
                className="ds-c-button ds-c-button--small hc-c-action-menu-button ds-u-sm-margin-left--2"
                onClick={[Function]}
                type="button"
              >
                <ClearIconAlternate>
                  <svg
                    className="ds-c-clear-icon"
                    height="14"
                    viewBox="0 0 14 14"
                    width="14"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m8.6002309 7 5.2573243-5.25732429c.1899264-.1899264.1899264-.49704142 0-.68696782l-.9132631-.91326309c-.1899264-.1899264-.4970414-.1899264-.6869678 0l-5.2573243 5.2573243-5.25732429-5.2573243c-.1899264-.1899264-.49704142-.1899264-.68696782 0l-.91326309.91326309c-.1899264.1899264-.1899264.49704142 0 .68696782l5.2573243 5.25732429-5.2573243 5.2573243c-.1899264.1899264-.1899264.4970414 0 .6869678l.91326309.9132631c.1899264.1899264.49704142.1899264.68696782 0l5.25732429-5.2573243 5.2573243 5.2573243c.1899264.1899264.4970414.1899264.6869678 0l.9132631-.9132631c.1899264-.1899264.1899264-.4970414 0-.6869678z"
                    />
                  </svg>
                </ClearIconAlternate>
                header.menu
              </button>
            </Button>
          </MenuButton>
        </div>
      </LoggedInActionMenu>
    </nav>
  </ActionMenu>
</Component>
`;

exports[`ActionMenu logged-out version renders logged-out version 1`] = `
<Component
  links={
    Array [
      Object {
        "href": "href",
        "label": "label",
      },
    ]
  }
  locale="en"
  loggedIn={false}
  onMenuToggleClick={[MockFunction]}
>
  <ActionMenu
    links={
      Array [
        Object {
          "href": "href",
          "label": "label",
        },
      ]
    }
    locale="en"
    loggedIn={false}
    onMenuToggleClick={[MockFunction]}
    t={[Function]}
  >
    <nav
      className="hc-c-header__actions ds-l-col ds-l-col--auto ds-u-margin-left--auto ds-u-font-weight--bold"
      id="hc-c-header__actions"
      role="navigation"
    >
      <LoggedOutActionMenu
        links={
          Array [
            Object {
              "href": "href",
              "label": "label",
            },
          ]
        }
        locale="en"
        loggedIn={false}
        onMenuToggleClick={[MockFunction]}
        t={[Function]}
      >
        <div>
          <ul
            className="hc-c-logged-out-links ds-c-list--bare ds-u-display--none ds-u-sm-display--inline-block"
          >
            <li
              className="hc-c-logged-out-links__li"
              key="href"
            >
              <a
                className="hc-c-logged-out-links__link"
                href="href"
                onClick={[Function]}
              >
                label
              </a>
            </li>
          </ul>
          <MenuButton
            className="ds-u-display--inline-block ds-u-sm-display--none"
            links={
              Array [
                Object {
                  "href": "href",
                  "label": "label",
                },
              ]
            }
            locale="en"
            loggedIn={false}
            onMenuToggleClick={[MockFunction]}
            t={[Function]}
          >
            <Button
              aria-controls="hc-c-menu"
              aria-expanded={false}
              aria-label="header.openMenu"
              className="hc-c-action-menu-button ds-u-sm-margin-left--2 ds-u-display--inline-block ds-u-sm-display--none"
              component="button"
              onClick={[Function]}
              size="small"
              type="button"
            >
              <button
                aria-controls="hc-c-menu"
                aria-expanded={false}
                aria-label="header.openMenu"
                className="ds-c-button ds-c-button--small hc-c-action-menu-button ds-u-sm-margin-left--2 ds-u-display--inline-block ds-u-sm-display--none"
                onClick={[Function]}
                type="button"
              >
                <MenuIcon
                  className="ds-u-margin-right--1"
                >
                  <svg
                    aria-hidden="true"
                    className="svg-inline--fa fa-bars fa-w-14 ds-u-margin-right--1"
                    data-icon="bars"
                    data-prefix="fas"
                    viewBox="0 0 448 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
                      fill="currentColor"
                    />
                  </svg>
                </MenuIcon>
                header.menu
              </button>
            </Button>
          </MenuButton>
        </div>
      </LoggedOutActionMenu>
    </nav>
  </ActionMenu>
</Component>
`;
